CMakeLists.txt

```diff
@@ -886,6 +886,16 @@ cmake_dependent_option(
   Will be disabled if not supported by the platform" ON
   "USE_CUDA OR USE_ROCM" OFF)
 
+#
+# Cannot be put into Dependencies.cmake due circular dependency:
```

Comment:
Can you please elaborate what do you mean here?

Reply:
`USE_FLASH_ATTENTION` and `USE_MEM_EFF_ATTENTION` (defined above these lines) depend on the final value of USE_ROCM/USE_CUDA, which is only set after Dependencies.cmake is processed. Thus the inclusion of `aotriton.cmake` cannot be done inside Dependencies.cmake.

Reply:
Well this will be a little long...

For current build system, all `USE_XXX` variables serve quadruple roles.
1. "CMake CACHE entry", picked up by `setup.py` from env vars and send to cmake as `-D`
2. Also CMake CACHE entry, but supplied by CMakeCache.txt from previous build and used in incremental build.
3. Default values from `option` or `cmake_dependent_option`
4. Actual value used by late part of `CMakeLists.txt`, after being settled through some logic

For example, `USE_FLASH_ATTENTION` being `ON` could be any of these cases
1. Supplied by user as `USE_FLASH_ATTENTION=1`
2. Set by `cmake_dependent_option` as default value, because it's a CUDA/ROCM
3. Loaded from CMakeCache.txt since this is a incremental build

For 1 and 3, `USE_FLASH_ATTENTION` has reliable value from the first line of `CMakeLists.txt`, but Case 2 is a little different, because it requires `USE_CUDA` and/or `USE_ROCM`. This is the `USE_FLASH_ATTENTION -> USE_ROCM` part.

Similar to USE_FLASH_ATTENTION, USE_CUDA/USE_ROCM also have multiple sources, and only after `include(cmake/Dependencies.cmake)`, these values are reliable, because they are both default to ON on Linux, and supposed to be turned off accordingly in `Dependencies.cmake`. This is the `USE_ROCM -> Dependencies.cmake` part.

The last part `Dependencies.cmake -> aotriton.cmake` describes the status quo: `aotriton.cmake` is included in `Dependencies.cmake`, but `USE_FLASH_ATTENTION` does not have reliable value at that point, and hence the decision cannot be made in `Dependencies.cmake` and should be deferred to the main `CMakeLists.txt`, after `cmake_dependent_option(USE_FLASH_ATTENTION)`